src/zenml/pipelines/base_pipeline.py

```diff
@@ -158,9 +159,18 @@ def _verify_arguments(self, *steps: BaseStep, **kw_steps: BaseStep) -> None:
                     f"a pipeline."
                 )
 
+            step_ref_func = type(step)
+
+            if step_ref_func in step_refs:
+                raise PipelineInterfaceError(
+                    f"Step object (`{type(step)}`) has been used twice. Step "
+                    f"objects should be unique for each argument."
+                )
+
             key = input_step_keys[i]
             step.pipeline_parameter_name = key
             combined_steps[key] = step
+            step_refs[step_ref_func] = str(i)
```

Comment:
Do we need to coerce the `i` to a string?

Reply:
Not really I wanted to do it since it was a string as suggested below I will move it to a set rather than a Dict.